parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java

```diff
@@ -55,203 +73,239 @@
  *
  * @author Tianshuo Deng
  */
-public class ThriftSchemaConvertVisitor implements ThriftType.TypeVisitor {
+public class ThriftSchemaConvertVisitor implements ThriftType.TypeVisitor<ConvertedField, ThriftSchemaConvertVisitor.State> {
   private final FieldProjectionFilter fieldProjectionFilter;
-  private final FieldsPath currentFieldPath = new FieldsPath();
 
-  private Type currentType;
-  private Type.Repetition currentRepetition = Type.Repetition.REPEATED; // MessageType is repeated GroupType
-  private String currentName = "ParquetSchema";
+  private ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter) {
+    this.fieldProjectionFilter = checkNotNull(fieldProjectionFilter, "fieldProjectionFilter");
+  }
+
+  public static MessageType convert(StructType struct, FieldProjectionFilter filter) {
+    State state = new State(new FieldsPath(), REPEATED, "ParquetSchema");
+
+    ConvertedField conv = struct.accept(new ThriftSchemaConvertVisitor(filter), state);
+
+    MessageType messageType;
+
+    if (!conv.isKeep()) {
+      // TODO: should this be fatal? Projecting away everything?
+      messageType = new MessageType(state.name, new ArrayList<Type>());
+    } else {
+      messageType = new MessageType(state.name, conv.asKeep().getType().asGroupType().getFields());
+    }
 
-  public ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter) {
-    this.fieldProjectionFilter = Preconditions.checkNotNull(fieldProjectionFilter, "fieldProjectionFilter");
+    return messageType;
   }
 
   @Override
-  public void visit(ThriftType.MapType mapType) {
-    final ThriftField mapKeyField = mapType.getKey();
-    final ThriftField mapValueField = mapType.getValue();
-
-    //save env for map
-    String mapName = currentName;
-    Type.Repetition mapRepetition = currentRepetition;
-
-    //=========handle key
-    currentFieldPath.push(mapKeyField);
-    currentName = "key";
-    currentRepetition = REQUIRED;
-    mapKeyField.getType().accept(this);
-    Type keyType = currentType;//currentType is the already converted type
-    currentFieldPath.pop();
-
-    //=========handle value
-    currentFieldPath.push(mapValueField);
-    currentName = "value";
-    currentRepetition = OPTIONAL;
-    mapValueField.getType().accept(this);
-    Type valueType = currentType;
-    currentFieldPath.pop();
-
-    if (keyType == null && valueType == null) {
-      currentType = null;
-      return;
+  public ConvertedField visit(MapType mapType, State state) {
+    ThriftField keyField = mapType.getKey();
+    ThriftField valueField = mapType.getValue();
+
+    ConvertedField convKey = keyField.getType().accept(this,
+        new State(state.path.push(keyField), REQUIRED, "key"));
+
+    ConvertedField convValue = valueField.getType().accept(this,
+        new State(state.path.push(valueField), OPTIONAL, "value"));
+
+    if (!convKey.isKeep()) {
```

Comment:
Note: add tests for `Map<Union, T>`, `Map<T, Union>`, `Map<Union, Union>`, `List<Union>` etc.